src/rpc/claimtrie.cpp

```diff
@@ -35,12 +33,43 @@ static CBlockIndex* BlockHashIndex(const uint256& blockHash)
     if (!chainActive.Contains(pblockIndex))
         throw JSONRPCError(RPC_INTERNAL_ERROR, "Block not in main chain");
 
+    const size_t currentMemoryUsage = pcoinsTip->DynamicMemoryUsage();
+    const int MAX_RPC_BLOCK_DECREMENTS = nCoinCacheUsage / currentMemoryUsage;
+
     if (chainActive.Tip()->nHeight > (pblockIndex->nHeight + MAX_RPC_BLOCK_DECREMENTS))
         throw JSONRPCError(RPC_INTERNAL_ERROR, "Block is too deep");
 
     return pblockIndex;
 }
 
+bool RollBackTo(const CBlockIndex* targetIndex, CCoinsViewCache& coinsCache, CClaimTrieCache& trieCache)
+{
+    AssertLockHeld(cs_main);
+
+    const size_t currentMemoryUsage = pcoinsTip->DynamicMemoryUsage();
+    for (CBlockIndex* index = chainActive.Tip(); index && index != targetIndex; index = index->pprev) {
+        boost::this_thread::interruption_point();
+
+        CBlock block;
+        if (!ReadBlockFromDisk(block, index, Params().GetConsensus()))
+            throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf("Failed to read %s", index->ToString()));
+
+        if (coinsCache.DynamicMemoryUsage() + currentMemoryUsage > nCoinCacheUsage)
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Out of memory, you may want to increase dbcache size");
+
+        if (ShutdownRequested())
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Shutdown requested");
+
+        CValidationState state;
+        if (!DisconnectBlock(block, state, index, coinsCache, trieCache))
+            throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf("Failed to disconnect %s", block.ToString()));
+
+        if (state.IsError())
+            return false;
```

Comment:
Why not throw here as well? and make the method void return?

Reply:
Looks like state is totally unused in DisconnectBlock i'll remove that check entirely.